core/finality-grandpa/src/lib.rs

```diff
@@ -644,41 +853,129 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
 	{
 		use authorities::PendingChange;
 
+		// we don't want to finalize on `inner.import_block`
+		let justification = block.justification.take();
```

Comment:
does this comment mean that `import_block` automatically finalizes when `justification` is `Some`? 

The cloning and hashing within this `import_block` implementation could add up but perhaps this is premature optimization.

Reply:
Yes, that is the assumption. It only actually finalizes if `finalized = true`, but the contract of `import_block` has changed to: if a justification is provided, then the block must be finalized (this should maybe be statically typed instead of using different parameters for finalized and the justification).